drivers/i2c/i2c_tca954x.c

```diff
@@ -0,0 +1,208 @@
+{
+	struct tca954x_root_data *i2c_tca954x = dev->data;
+	const struct tca954x_root_config *config = dev->config;
+
+	if (!device_is_ready(config->bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus->name);
+		return -ENODEV;
+	}
+
+	/* If the RESET line is available, configure it. */
+	if (config->reset_gpios.port) {
+		if (!device_is_ready(config->reset_gpios.port)) {
+			LOG_ERR("%s is not ready",
+				config->reset_gpios.port->name);
+			return -ENODEV;
+		}
+
+		if (gpio_pin_configure_dt(&config->reset_gpios, GPIO_OUTPUT)) {
+			LOG_ERR("%s: failed to configure RESET line", dev->name);
+			return -EIO;
+		}
+
+		/* Deassert reset line */
+		gpio_pin_set_dt(&config->reset_gpios, 0);
+	}
+
+	i2c_tca954x->selected_chan = 0;
+
+	return 0;
+}
+
+static int tca954x_channel_init(const struct device *dev)
+{
+	const struct tca954x_channel_config *chan_cfg = dev->config;
+	const struct tca954x_root_config *root_cfg =
+			get_root_config_from_channel(dev);
+
+	if (!device_is_ready(chan_cfg->root)) {
+		LOG_ERR("I2C mux root %s not ready", chan_cfg->root->name);
+		return -ENODEV;
+	}
+
+	if (chan_cfg->chan_mask >= BIT(root_cfg->nchans)) {
+		LOG_ERR("Wrong DTS address provided for %s", dev->name);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+const struct i2c_driver_api tca954x_api_funcs = {
+	.configure = tca954x_configure,
+	.transfer = tca954x_transfer,
+};
+
+#define TCA954x_CHILD_DEFINE(node_id)					    \
+	static const struct tca954x_channel_config			    \
+		tca954x_down_config_##node_id = {			    \
+		.chan_mask = BIT(DT_REG_ADDR(node_id)),			    \
+		.root = DEVICE_DT_GET(DT_PARENT(node_id)),		    \
+	};								    \
+	DEVICE_DT_DEFINE(node_id,					    \
+			 tca954x_channel_init,				    \
+			 NULL,						    \
+			 NULL,						    \
+			 &tca954x_down_config_##node_id,		    \
+			 POST_KERNEL, CONFIG_I2C_TCA954X_CHANNEL_INIT_PRIO, \
+			 &tca954x_api_funcs);
+
+#define TCA954x_ROOT_DEFINE(inst, ch, compat)				          \
+	static const struct tca954x_root_config tca954x_cfg_##inst = {            \
+		.slave_addr = DT_REG_ADDR_BY_IDX(DT_INST(inst, compat), 0),       \
+		.bus = DEVICE_DT_GET(DT_BUS(DT_INST(inst, compat))),		  \
+		.nchans = ch,			                                  \
+		.reset_gpios = GPIO_DT_SPEC_GET_OR(			          \
+					DT_INST(inst, compat), reset_gpios, {0}), \
+	};								          \
+	static struct tca954x_root_data tca954x_data_##inst = {	      \
+		.lock = Z_MUTEX_INITIALIZER(tca954x_data_##inst.lock),	      \
+	};								      \
+	DEVICE_DT_DEFINE(DT_INST(inst, compat),					      \
+			      tca954x_root_init, NULL,			      \
+			      &tca954x_data_##inst, &tca954x_cfg_##inst,    \
+			      POST_KERNEL, CONFIG_I2C_TCA954X_ROOT_INIT_PRIO, \
+			      NULL);					      \
+	DT_FOREACH_CHILD(DT_INST(inst, compat), TCA954x_CHILD_DEFINE);
+
+#define TCA954X_FOREACH_STATUS_OKAY(compat, fn)		\
+	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),		\
+		    (UTIL_CAT(DT_FOREACH_OKAY_INST_,		\
+			      compat)(fn)),			\
```

Comment:
You are using the generated macros directly here; this is not allowed. These are an implementation detail that should not be used directly from a driver. Please use DT_INST_FOREACH_STATUS_OKAY instead.

Reply:
I think that replacing `TCA954X_FOREACH_STATUS_OKAY` by `DT_FOREACH_STATUS_OKAY` is the cleanest way to do it with DT macros.
Being based on the adc lmp90xxx driver, do you want me to push another PR to fix this in this one too ?

Reply:
After testing, it wont work since `DT_FOREACH_STATUS_OKAY` is declaring `DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_S_mux_i2c_2_REG_IDX_0_VAL_ADDRESS` while my calls for `DT_INST` are searching for `DT_N_INST_DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_ti_tca9548a_REG_IDX_0_VAL_ADDRESS`.

So if `DT_INST_FOREACH_STATUS_OKAY` cannot take a compat as argument and there isn't an equivalent to `DT_INST` that is `UTIL_CAT(DT_N_, DT_DASH(inst, compat))` the way I did it is the only way I see to do it !

Reply:
After some testing, it's not possible to use `DT_INST_FOREACH_STATUS_OKAY` because it wont take compat as argument and we need it as soon as we support multiple compatible.
I tried using `DT_FOREACH_STATUS_OKAY` but it wont work either because `DT_INST` is then searching for a `DT_INST_FOREACH_STATUS_OKAY` generated field.

That's why we use `TCA954X_FOREACH_STATUS_OKAY` to do `DT_INST_FOREACH_STATUS_OKAY` that can take a compat as argument.
I tried multiple ways to do it but can't find any that is working except the one already used.

Reply:
You can `#undef DT_DRV_COMPAT` and then define it to something else before using `DT_INST_FOREACH_STATUS_OKAY`. See, for example, drivers/serial/uart_gecko.c.

Reply:
Just tested this :
```
/*
 * TCA9546A: 4 channels
 */
#define TCA9546A_INIT(n) TCA954x_ROOT_DEFINE(n, 4, ti_tca9546a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9546a
DT_INST_FOREACH_STATUS_OKAY(TCA9546A_INIT)

/*
 * TCA9548A: 8 channels
 */
#define TCA9548A_INIT(n) TCA954x_ROOT_DEFINE(n, 8, ti_tca9548a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9548a
DT_INST_FOREACH_STATUS_OKAY(TCA9548A_INIT)
```

It's working, is there a way to do it in a cleaner way or it is what you were thinking of @mbolivar-nordic ? Thanks

Reply:
fixed it